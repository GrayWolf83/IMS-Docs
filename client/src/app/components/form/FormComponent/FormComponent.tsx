import { Box, Button, Flex, useColorMode, useToast } from '@chakra-ui/react'
import React from 'react'

interface IFormComponent {
	initialData?: { [key: string]: string }
	onSubmit?: (data: { [key: string]: string | File }) => void
	onSubmitFile?: (data: FormData) => void
	btnLabel: string
	children: React.ReactNode
	validationShema: any
}

const FormComponent: React.FC<IFormComponent> = ({
	initialData,
	onSubmit,
	onSubmitFile,
	btnLabel,
	children,
	validationShema,
}) => {
	const [data, setData] = React.useState<{
		[key: string]: string | File
	}>({})
	const [error, setError] = React.useState<{ [key: string]: string }>({})
	const { colorMode } = useColorMode()
	const toast = useToast()

	React.useEffect(() => {
		if (initialData) {
			setData(initialData)
		}

		return () => setData({})
	}, [initialData])

	const validation = React.useCallback(() => {
		validationShema
			.validate(data)
			.then(() => setError({}))
			.catch((err: { message: { name: string; text: string } }) =>
				setError({ [err.message.name]: err.message.text }),
			)
	}, [validationShema, data])

	React.useEffect(() => {
		if (Object.keys(data).length) validation()
	}, [data, validation])

	const changeHandler = (value: { [key: string]: string | File }) => {
		setData((prevState) => ({
			...prevState,
			...value,
		}))
	}

	const submitHandler = (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault()
		validation()
		delete error?.undefined
		if (!Object.keys(error).length && Object.keys(data).length) {
			if (onSubmitFile) {
				const formData = new FormData()

				Object.entries(data).forEach((item) => {
					formData.append(item[0], item[1])
				})

				onSubmitFile(formData)
			}

			if (onSubmit) {
				onSubmit(data)
			}

			setData({})
		} else {
			toast({
				title: 'Заполните все поля формы',
				status: 'error',
				variant: 'solid',
				isClosable: true,
			})
		}
	}

	return (
		<Box
			data-testid='FormComponent'
			width={['95%', '75%', '50%', '40%']}
			mx='auto'
			bg={colorMode === 'dark' ? 'gray.600' : 'gray.300'}
			mt='50px'
			p='20px'
			borderRadius='10px'
			shadow='md'>
			<form
				onSubmit={submitHandler}
				encType='multipart/form-data'
				style={{ width: '100%' }}>
				{React.Children.map(children, (child: any) => {
					const config = {
						...child.props,
						onChange: changeHandler,
						value: data[child.props.name] || '',
						error: error[child?.props?.name],
					}

					return React.cloneElement(child, config)
				})}

				<Flex justifyContent='center'>
					<Button my='10px' colorScheme='gray' type='submit'>
						{btnLabel}
					</Button>
				</Flex>
			</form>
		</Box>
	)
}

export default FormComponent
