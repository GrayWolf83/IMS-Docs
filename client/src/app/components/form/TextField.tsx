import React from 'react'
import {
	FormControl,
	FormLabel,
	Input,
	useColorMode,
	Flex,
	IconButton,
	FormErrorMessage,
} from '@chakra-ui/react'
import { FaEye, FaEyeSlash } from 'react-icons/fa'

interface IProps {
	error?: string | null
	label: string
	name: string
	type?: string
	onChange?: (value: { [key: string]: string }) => void
	value?: string
}

const TextField: React.FC<IProps> = ({
	name,
	error,
	label,
	type = 'text',
	onChange,
	value,
}) => {
	const { colorMode } = useColorMode()
	const [passVisible, setPassVisible] = React.useState(false)

	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		if (onChange) {
			onChange({ [e.target.name]: e.target.value })
		}
	}

	return (
		<FormControl isRequired my='20px' isInvalid={Boolean(error)}>
			<FormLabel>{label}</FormLabel>
			<Flex>
				<Input
					autoComplete={name}
					name={name}
					borderColor={colorMode === 'dark' ? 'gray.300' : 'gray.600'}
					_focusVisible={{
						borderColor:
							colorMode === 'dark' ? 'gray.300' : 'gray.600',
					}}
					placeholder='Пароль'
					onChange={handleChange}
					value={value}
					type={
						type === 'password'
							? passVisible
								? 'text'
								: 'password'
							: type
					}
				/>
				{type === 'password' && (
					<IconButton
						size='md'
						fontSize='lg'
						variant='outline'
						borderColor={
							colorMode === 'dark' ? 'gray.300' : 'gray.600'
						}
						color='current'
						marginLeft='1'
						aria-label=''
						onClick={() => setPassVisible(!passVisible)}
						icon={
							passVisible ? (
								<FaEyeSlash fontSize='24px' />
							) : (
								<FaEye fontSize='24px' />
							)
						}
					/>
				)}
			</Flex>
			{error && <FormErrorMessage>{error}</FormErrorMessage>}
		</FormControl>
	)
}

export default TextField
