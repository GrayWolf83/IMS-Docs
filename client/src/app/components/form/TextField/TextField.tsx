import React from 'react'
import {
	FormControl,
	FormLabel,
	Input,
	Flex,
	IconButton,
	FormErrorMessage,
} from '@chakra-ui/react'
import { FaEye, FaEyeSlash } from 'react-icons/fa'
import useColor from '../../../hooks/useColor'

interface IProps {
	error?: string | null
	label: string
	name: string
	type?: string
	onChange?: (value: { [key: string]: string }) => void
	value?: string
	placeholder: string
}

const TextField: React.FC<IProps> = ({
	name,
	error,
	label,
	type = 'text',
	onChange,
	value,
	placeholder,
}) => {
	const { light } = useColor()
	const [passVisible, setPassVisible] = React.useState(false)

	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		if (onChange) {
			onChange({ [e.target.name]: e.target.value })
		}
	}

	return (
		<FormControl
			data-testid='TextField'
			my='20px'
			isInvalid={Boolean(error)}>
			<FormLabel>{label}</FormLabel>
			<Flex>
				<Input
					autoComplete={name}
					name={name}
					borderColor={light}
					_focusVisible={{
						borderColor: 'none',
					}}
					_hover={{
						borderColor: light,
					}}
					placeholder={placeholder}
					onChange={handleChange}
					value={value}
					type={
						type === 'password'
							? passVisible
								? 'text'
								: 'password'
							: type
					}
				/>
				{type === 'password' && (
					<IconButton
						size='md'
						fontSize='lg'
						variant='outline'
						borderColor={light}
						color='current'
						marginLeft='1'
						aria-label=''
						onClick={() => setPassVisible(!passVisible)}
						icon={
							passVisible ? (
								<FaEyeSlash fontSize='24px' />
							) : (
								<FaEye fontSize='24px' />
							)
						}
					/>
				)}
			</Flex>
			{error && <FormErrorMessage>{error}</FormErrorMessage>}
		</FormControl>
	)
}

export default TextField
