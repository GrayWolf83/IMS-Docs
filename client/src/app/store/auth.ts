import { setLoadingError } from './error'
import { IUser } from './../../../../server/types/User'
import { AppDispatch, RootState } from './index'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import authService from '../services/auth.service'
import localStorageService from '../services/localStorage.service'

type authState = {
	currentUser: IUser | null
	isLoading: boolean
}

const initialState: authState = {
	currentUser: null,
	isLoading: false,
}

const authSlice = createSlice({
	name: 'auth',
	initialState,
	reducers: {
		authLoginSuccess(state, action: PayloadAction<IUser>) {
			state.currentUser = action.payload
		},
		authLogout(state) {
			state.currentUser = null
		},
		authLoadingStart(state) {
			state.isLoading = true
		},
		authLoadingEnd(state) {
			state.isLoading = false
		},
	},
})

const { authLoginSuccess, authLoadingStart, authLoadingEnd, authLogout } =
	authSlice.actions

export const login = (data: FormData) => async (dispatch: AppDispatch) => {
	dispatch(authLoadingStart())

	try {
		const payload = await authService.login(data)
		dispatch(authLoginSuccess(payload.user))

		localStorageService.setTokens({
			...payload,
		})
	} catch (error) {
		dispatch(setLoadingError(error))
	} finally {
		dispatch(authLoadingEnd())
	}
}

export const autologin = () => async (dispatch: AppDispatch) => {
	dispatch(authLoadingStart())
	try {
		const payload = await authService.refresh()

		if (!payload?.refreshToken || !payload?.accessToken) {
			localStorageService.removeAuthData()
		} else {
			dispatch(authLoginSuccess(payload.user))

			localStorageService.setTokens({
				...payload,
			})
		}
	} catch (error: any) {
		dispatch(setLoadingError(error))
	} finally {
		dispatch(authLoadingEnd())
	}
}

export const logout = () => async (dispatch: AppDispatch) => {
	localStorageService.removeAuthData()
	dispatch(authLogout())
}

export const getCurrentUser = () => (state: RootState) => {
	return state.auth.currentUser
}

export const getAuthLoadingStatus = () => (state: RootState) => {
	return state.auth.isLoading
}

export const getCurrentUserIsManage = () => (state: RootState) => {
	return Boolean(state.auth.currentUser?.role === 'Manage')
}

export default authSlice.reducer
