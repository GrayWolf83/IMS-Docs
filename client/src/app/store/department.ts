import { setLoadingError } from './error'
import { AppDispatch, RootState } from './index'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IDepartment } from '../types/department'
import departmentService from '../services/department.service'

type state = {
	entities: IDepartment[]
	isLoading: boolean
}

const initialState: state = {
	entities: [],
	isLoading: false,
}

const departmentSlice = createSlice({
	name: 'departments',
	initialState,
	reducers: {
		departmentsLoaded(state, action: PayloadAction<IDepartment[]>) {
			state.entities = action.payload
		},
		departmentAdded(state, action: PayloadAction<IDepartment>) {
			state.entities = [action.payload, ...state.entities]
		},
		departmentEdited(state, action: PayloadAction<IDepartment>) {
			state.entities = state.entities.map((item) => {
				return item.id === action.payload.id ? action.payload : item
			})
		},
		departmentDeleted(state, action: PayloadAction<string>) {
			state.entities = state.entities.filter(
				(item) => item.id !== action.payload,
			)
		},
		departmentLoadingStart(state) {
			state.isLoading = true
		},
		departmentLoadingEnd(state) {
			state.isLoading = false
		},
	},
})

const {
	departmentsLoaded,
	departmentAdded,
	departmentEdited,
	departmentDeleted,
	departmentLoadingStart,
	departmentLoadingEnd,
} = departmentSlice.actions

export const loadDepartmentList = () => async (dispatch: AppDispatch) => {
	dispatch(departmentLoadingStart())

	try {
		const payload = await departmentService.getList()
		dispatch(departmentsLoaded(payload))
	} catch (error) {
		dispatch(setLoadingError(error))
	} finally {
		dispatch(departmentLoadingEnd())
	}
}

export const addDepartment =
	(data: { name: string; fullName: string; index: number }) =>
	async (dispatch: AppDispatch) => {
		dispatch(departmentLoadingStart())
		try {
			const payload: IDepartment = await departmentService.add(data)

			dispatch(departmentAdded(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(departmentLoadingEnd())
		}
	}

export const editDepartment =
	(data: IDepartment) => async (dispatch: AppDispatch) => {
		dispatch(departmentLoadingStart())
		try {
			const payload: IDepartment = await departmentService.edit(data)

			dispatch(departmentEdited(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(departmentLoadingEnd())
		}
	}

export const deleteDepartment =
	(data: string) => async (dispatch: AppDispatch) => {
		dispatch(departmentLoadingStart())
		try {
			const payload: string = await departmentService.delete(data)

			dispatch(departmentDeleted(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(departmentLoadingEnd())
		}
	}

export const getDepartmentsList = () => (state: RootState) => {
	return state.departments.entities
		.filter((item) => item.index)
		.sort((a, b) => (a.index > b.index ? 1 : -1))
}

export const getDepartmentById = (id: string) => (state: RootState) => {
	return state.departments.entities.find((item) => item.id === id)
}

export default departmentSlice.reducer
