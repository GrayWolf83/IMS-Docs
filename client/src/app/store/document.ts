import { setLoadingError } from './error'
import { AppDispatch, RootState } from './index'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IDocument, INewDocument } from '../types/document'
import documentService from '../services/document.service'

type state = {
	entities: IDocument[]
	isLoading: boolean
}

const initialState: state = {
	entities: [],
	isLoading: false,
}

const documentSlice = createSlice({
	name: 'documents',
	initialState,
	reducers: {
		loaded(state, action: PayloadAction<IDocument[]>) {
			state.entities = action.payload
		},
		added(state, action: PayloadAction<IDocument>) {
			state.entities = [action.payload, ...state.entities]
		},
		edited(state, action: PayloadAction<IDocument>) {
			state.entities = state.entities.map((item) => {
				return item.id === action.payload.id ? action.payload : item
			})
		},
		deleted(state, action: PayloadAction<string>) {
			state.entities = state.entities.filter(
				(item) => item.id !== action.payload,
			)
		},
		loadingStart(state) {
			state.isLoading = true
		},
		loadingEnd(state) {
			state.isLoading = false
		},
	},
})

const { loaded, added, edited, deleted, loadingStart, loadingEnd } =
	documentSlice.actions

export const loadDocumentsList =
	(systemId: string) => async (dispatch: AppDispatch) => {
		dispatch(loadingStart())

		try {
			const payload: IDocument[] = await documentService.getList(systemId)
			dispatch(loaded(payload))
		} catch (error) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(loadingEnd())
		}
	}

export const addDocument =
	(data: INewDocument) => async (dispatch: AppDispatch) => {
		dispatch(loadingStart())
		try {
			const payload: IDocument = await documentService.add(data)

			dispatch(added(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(loadingEnd())
		}
	}

export const editDocument =
	(data: IDocument) => async (dispatch: AppDispatch) => {
		dispatch(loadingStart())
		try {
			const payload: IDocument = await documentService.edit(data)

			dispatch(edited(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(loadingEnd())
		}
	}

export const deleteDocument =
	(data: string) => async (dispatch: AppDispatch) => {
		dispatch(loadingStart())
		try {
			const payload: string = await documentService.delete(data)

			dispatch(deleted(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(loadingEnd())
		}
	}

export const getDocumentsList = () => (state: RootState) => {
	return state.documents.entities
}

export const getDocumentById = (id: string) => (state: RootState) => {
	return state.documents.entities.find((item) => item.id === id)
}

export default documentSlice.reducer
