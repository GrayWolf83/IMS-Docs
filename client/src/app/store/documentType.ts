import { setLoadingError } from './error'
import { AppDispatch, RootState } from './index'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IDocumentType, INewDocumentType } from '../types/documentType'
import documentTypeService from '../services/documentType.service'

type state = {
	entities: IDocumentType[]
	isLoading: boolean
}

const initialState: state = {
	entities: [],
	isLoading: false,
}

const documentTypeSlice = createSlice({
	name: 'documentTypes',
	initialState,
	reducers: {
		documentTypesLoaded(state, action: PayloadAction<IDocumentType[]>) {
			state.entities = action.payload
		},
		documentTypeAdded(state, action: PayloadAction<IDocumentType>) {
			state.entities = [action.payload, ...state.entities]
		},
		documentTypeEdited(state, action: PayloadAction<IDocumentType>) {
			state.entities = state.entities.map((item) => {
				return item.id === action.payload.id ? action.payload : item
			})
		},
		documentTypeDeleted(state, action: PayloadAction<string>) {
			state.entities = state.entities.filter(
				(item) => item.id !== action.payload,
			)
		},
		documentTypeLoadingStart(state) {
			state.isLoading = true
		},
		documentTypeLoadingEnd(state) {
			state.isLoading = false
		},
	},
})

const {
	documentTypesLoaded,
	documentTypeAdded,
	documentTypeEdited,
	documentTypeDeleted,
	documentTypeLoadingStart,
	documentTypeLoadingEnd,
} = documentTypeSlice.actions

export const loadDocumentTypesList = () => async (dispatch: AppDispatch) => {
	dispatch(documentTypeLoadingStart())

	try {
		const payload: IDocumentType[] = await documentTypeService.getList()
		dispatch(documentTypesLoaded(payload))
	} catch (error) {
		dispatch(setLoadingError(error))
	} finally {
		dispatch(documentTypeLoadingEnd())
	}
}

export const addDocumentType =
	(data: INewDocumentType) => async (dispatch: AppDispatch) => {
		dispatch(documentTypeLoadingStart())
		try {
			const payload: IDocumentType = await documentTypeService.add(data)

			dispatch(documentTypeAdded(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(documentTypeLoadingEnd())
		}
	}

export const editDocumentType =
	(data: IDocumentType) => async (dispatch: AppDispatch) => {
		dispatch(documentTypeLoadingStart())
		try {
			const payload: IDocumentType = await documentTypeService.edit(data)

			dispatch(documentTypeEdited(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(documentTypeLoadingEnd())
		}
	}

export const deleteDocumentType =
	(data: string) => async (dispatch: AppDispatch) => {
		dispatch(documentTypeLoadingStart())
		try {
			const payload: string = await documentTypeService.delete(data)

			dispatch(documentTypeDeleted(payload))
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(documentTypeLoadingEnd())
		}
	}

export const getDocumentTypesList = () => (state: RootState) => {
	return state.documentTypes.entities
}

export const getDocumentTypeById = (id: string) => (state: RootState) => {
	return state.documentTypes.entities.find((item) => item.id === id)
}

export default documentTypeSlice.reducer
