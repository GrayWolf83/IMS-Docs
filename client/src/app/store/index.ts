import { configureStore } from '@reduxjs/toolkit'
import errorReducer from './error'
import authReducer from './auth'
import systemReducer from './system'
import departmentReducer from './department'

const store = configureStore({
	reducer: {
		errors: errorReducer,
		auth: authReducer,
		systems: systemReducer,
		departments: departmentReducer,
	},
})

export default store

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
