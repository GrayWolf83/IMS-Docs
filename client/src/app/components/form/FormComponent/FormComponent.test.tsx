import React from 'react'
import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from '../../../store'
import FormComponent from './FormComponent'
import TextField from '../TextField'

describe('FormComponent component', () => {
	test('FormComponent render', () => {
		render(
			<Provider store={store}>
				<BrowserRouter>
					<FormComponent
						btnLabel='Form'
						onSubmit={() => {}}
						validationShema={{}}>
						<TextField
							name='data'
							value=''
							onChange={() => {}}
							error={null}
							placeholder='test'
							label='data'
						/>
					</FormComponent>
				</BrowserRouter>
			</Provider>,
		)

		const formComponent = screen.getByTestId('FormComponent')
		expect(formComponent).toBeInTheDocument()
	})

	test('FormComponent snapshot', () => {
		expect(
			render(
				<Provider store={store}>
					<BrowserRouter>
						<FormComponent
							btnLabel='Form'
							onSubmit={() => {}}
							validationShema={{}}>
							<TextField
								name='data'
								value=''
								onChange={() => {}}
								error={null}
								placeholder='test'
								label='data'
							/>
						</FormComponent>
					</BrowserRouter>
				</Provider>,
			),
		).toMatchSnapshot()
	})
})
