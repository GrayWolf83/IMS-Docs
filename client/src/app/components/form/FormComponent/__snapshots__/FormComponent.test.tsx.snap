// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormComponent component FormComponent snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-0"
        data-testid="FormComponent"
      >
        <form
          enctype="multipart/form-data"
          style="width: 100%;"
        >
          <div
            class="chakra-form-control css-0"
            data-testid="TextField"
            role="group"
          >
            <label
              class="chakra-form__label css-0"
              for="field-:r1:"
              id="field-:r1:-label"
            >
              data
            </label>
            <div
              class="css-0"
            >
              <input
                autocomplete="data"
                class="chakra-input css-0"
                id="field-:r1:"
                name="data"
                placeholder="test"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="css-0"
          >
            <button
              class="chakra-button css-0"
              type="submit"
            >
              Form
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-0"
      data-testid="FormComponent"
    >
      <form
        enctype="multipart/form-data"
        style="width: 100%;"
      >
        <div
          class="chakra-form-control css-0"
          data-testid="TextField"
          role="group"
        >
          <label
            class="chakra-form__label css-0"
            for="field-:r1:"
            id="field-:r1:-label"
          >
            data
          </label>
          <div
            class="css-0"
          >
            <input
              autocomplete="data"
              class="chakra-input css-0"
              id="field-:r1:"
              name="data"
              placeholder="test"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="css-0"
        >
          <button
            class="chakra-button css-0"
            type="submit"
          >
            Form
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
